requirements_tools/upgrade_requirements.py

```diff
@@ -63,12 +69,14 @@ def installed(requirements_file):
         req = requirements_to_parse.pop()
         installed_req = installed_things[req.key]
         expected_pinned.add(
-            '{}=={}'.format(installed_req.project_name, installed_req.version),
+            f'{installed_req.project_name}=={installed_req.version}',
         )
         for sub in installed_req.requires(req.extras):
             if sub.key not in installed_things:
-                specifiers = ','.join(str(s) for s in sub.specifier)
-                unmet.add('{}{}'.format(sub.key, specifiers))
+                specifiers = ','.join(
+                    str(s) for s in sub.specifier  # type: ignore[attr-defined]
```

Comment:
the `Requirements` object is constructed a little strange (only via `.parse`) mypy doesn't get this...